libtransact/src/scheduler/serial/core.rs

```diff
@@ -156,92 +170,150 @@ impl SchedulerCore {
-                    // we panic as handling this is not currently implemented.
-                    unimplemented!("cannot handle ill-formed transaction: {}", err)
-                }
-            };
-
-            let context_id = match self.previous_context {
-                Some(previous_context_id) => self
-                    .context_lifecycle
-                    .create_context(&[previous_context_id], &self.state_id),
-                None => self.context_lifecycle.create_context(&[], &self.state_id),
-            };
-
-            self.current_batch = Some(pair);
-            self.execution_tx
-                .send(ExecutionTask::new(transaction, context_id))?
-        }
+        let context_id = match self.previous_context {
+            Some(previous_context_id) => self
+                .context_lifecycle
+                .create_context(&[previous_context_id], &self.state_id),
+            None => self.context_lifecycle.create_context(&[], &self.state_id),
+        };
+
+        self.current_txn = Some(transaction_pair.clone());
+        self.execution_tx
+            .send(ExecutionTask::new(transaction_pair, context_id))?;
+        self.next_ready = false;
 
         Ok(())
     }
 
+    fn invalidate_current_batch(&mut self, invalid_result: InvalidTransactionResult) {
+        // Invalidate all previously executed transactions in the batch
+        self.txn_results = self
+            .txn_results
+            .iter()
+            .map(|result| match result {
+                TransactionExecutionResult::Valid(receipt) => {
+                    TransactionExecutionResult::Invalid(InvalidTransactionResult {
+                        transaction_id: receipt.transaction_id.clone(),
+                        error_message: "batch invalid".into(),
+                        error_data: vec![],
+                    })
+                }
+                TransactionExecutionResult::Invalid(_) => {
+                    panic!("Should not already have an invalid result");
+                }
+            })
+            .collect();
+
+        self.txn_results
+            .push(TransactionExecutionResult::Invalid(invalid_result));
+
+        // Invalidate all unexecuted transactions in the batch
+        self.txn_results.append(
+            &mut self
+                .txn_queue
+                .drain(..)
+                .map(|txn| {
+                    TransactionExecutionResult::Invalid(InvalidTransactionResult {
+                        transaction_id: txn.header_signature().into(),
+                        error_message: "batch invalid".into(),
+                        error_data: vec![],
+                    })
+                })
+                .collect(),
+        );
+    }
+
+    fn send_batch_result(&mut self) {
+        let batch = self
+            .current_batch
+            .take()
+            .expect("received execution result but no current batch is executing");
```

Comment:
Can we restructure to get rid of some of the expect panics? Every one of these panics is an opportunity to crash the process (which isn't a nice attribute for library code). (In this case, I think it was pre-existing.)

Reply:
We _can_, but I'm not sure what the behavior should be in these scenarios. As far as I can tell, these are all things that indicate programmer error, which should be verified by testing.

Reply:
This returns an error now